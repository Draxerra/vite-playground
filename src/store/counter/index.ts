import { createAsyncThunk, createSlice, PayloadAction } from "@reduxjs/toolkit";

export const asyncIncrement = createAsyncThunk(
  "counter/asyncIncrement",
  async () => {
    await new Promise((resolve) => setTimeout(resolve, 5000));
    return 1;
  }
);

export const counter = createSlice({
  name: "counter",
  initialState: {
    value: 0,
  },
  reducers: {
    increment: (state) => {
      state.value += 1;
    },
    decrement: (state) => {
      state.value -= 1;
    },
    incrementByAmount: (state, action: PayloadAction<number>) => {
      state.value += action.payload;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(asyncIncrement.fulfilled, (state, action) => {
      state.value += action.payload;
    });
  },
});

export const { increment, decrement, incrementByAmount } = counter.actions;

export default counter.reducer;
