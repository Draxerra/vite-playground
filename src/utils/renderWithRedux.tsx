import React, { FC } from "react";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";
import { configureStore } from "@reduxjs/toolkit";

type RenderParams = Parameters<typeof render>;
type Render = ReturnType<typeof render>;
type StoreParams = Parameters<typeof configureStore>;
type Store = ReturnType<typeof configureStore>;

type RenderWithReduxOptions = {
  preloadedState?: StoreParams[0]["preloadedState"];
  reducers: StoreParams[0]["reducer"];
  store?: Store;
} & RenderParams[1];

const renderWithRedux: (
  ui: RenderParams[0],
  opts: RenderWithReduxOptions
) => Render = (
  ui,
  {
    preloadedState = {},
    reducers,
    store = configureStore({ reducer: reducers, preloadedState }),
    wrapper: WrapperElement,
    ...renderOptions
  }
) => {
  const Wrapper: FC = ({ children }) => (
    <Provider store={store}>
      {WrapperElement ? <WrapperElement>{children}</WrapperElement> : children}
    </Provider>
  );
  return render(ui, { wrapper: Wrapper, ...renderOptions });
};

export * from "@testing-library/react";
export { renderWithRedux };
