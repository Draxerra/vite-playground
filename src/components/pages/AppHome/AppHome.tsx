import React, { FC } from "react";

import { useAppDispatch, useAppSelector } from "~/store";
import { decrement, increment } from "~/store/counter";

import AppCounter from "~/components/molecules/AppCounter";

const AppHome: FC = () => {
  const count = useAppSelector((state) => state.counter.value);
  const dispatch = useAppDispatch();

  return (
    <AppCounter
      decrement={() => dispatch(decrement())}
      increment={() => dispatch(increment())}
    >
      {count}
    </AppCounter>
  );
};

export default AppHome;
