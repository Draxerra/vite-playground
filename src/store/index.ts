import { configureStore } from "@reduxjs/toolkit";
import { useDispatch, useSelector, TypedUseSelectorHook } from "react-redux";

import counter from "./counter";

const store = configureStore({
  reducer: {
    counter,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware(),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export const useAppDispatch = (): AppDispatch => useDispatch();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;

export default store;
