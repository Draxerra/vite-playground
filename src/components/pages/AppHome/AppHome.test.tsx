import React from "react";
import AppHome from "./AppHome";
import { screen, renderWithRedux } from "~/utils/renderWithRedux";
import userEvent from "@testing-library/user-event";

import counter from "~/store/counter";
import * as counterActions from "~/store/counter";

describe("App Home", () => {
  afterEach(() => {
    jest.restoreAllMocks();
  });

  it("should match snapshot", () => {
    const home = renderWithRedux(<AppHome />, {
      reducers: { counter },
    }).asFragment();
    expect(home).toMatchSnapshot();
  });

  it("should display count from store", () => {
    renderWithRedux(<AppHome />, {
      reducers: { counter },
      preloadedState: { counter: { value: 500 } },
    });
    expect(screen.getByText("Current Count: 500")).toBeInTheDocument();
  });

  it("should fire increment action", () => {
    renderWithRedux(<AppHome />, {
      reducers: { counter },
    });
    const spy = jest.spyOn(counterActions, "increment");
    const incrementBtn = screen.getByText("Increment");
    userEvent.click(incrementBtn);
    expect(spy).toHaveBeenCalled();
  });

  it("should increment on click", () => {
    renderWithRedux(<AppHome />, {
      reducers: { counter },
    });
    const incrementBtn = screen.getByText("Increment");
    userEvent.click(incrementBtn);
    expect(screen.getByText("Current Count: 1")).toBeInTheDocument();
  });

  it("should fire decrement action", () => {
    renderWithRedux(<AppHome />, {
      reducers: { counter },
    });
    const spy = jest.spyOn(counterActions, "decrement");
    const decrementBtn = screen.getByText("Decrement");
    userEvent.click(decrementBtn);
    expect(spy).toHaveBeenCalled();
  });

  it("should decrement on click", () => {
    renderWithRedux(<AppHome />, {
      reducers: { counter },
    });
    const decrementBtn = screen.getByText("Decrement");
    userEvent.click(decrementBtn);
    expect(screen.getByText("Current Count: -1")).toBeInTheDocument();
  });
});
