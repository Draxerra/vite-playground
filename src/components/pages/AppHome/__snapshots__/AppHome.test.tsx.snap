// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Home should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="appCounter"
  >
    <span
      aria-live="polite"
      class="appCounterText fs-600"
      role="region"
    >
      Current Count: 0
    </span>
    <button
      aria-label="Increment counter"
      class="appButton"
    >
      Increment
    </button>
    <button
      aria-label="Decrement counter"
      class="appButton"
    >
      Decrement
    </button>
  </div>
</DocumentFragment>
`;
